<a href="https://colab.research.google.com/github/aqmalio/Data-Science-2026/blob/main/Pertemuan3_Aqmal_250401020149.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

* Nama: Aqmal
* Nim: 250401020149
* Kelas: IF405

In [13]:
import pandas as pd, numpy as np
from scipy.stats.mstats import winsorize

# STEP 0 — Load & eksplorasi awal
df = pd.read_csv('https://raw.githubusercontent.com/aqmalio/Data-Science-2026/refs/heads/main/housing_dirty.csv')
print('Shape awal:', df.shape)
print(df.isnull().sum())

Shape awal: (130, 7)
id               0
luas_m2         18
harga_juta      17
kota             0
kamar           10
tahun_bangun     0
kondisi          0
dtype: int64


In [14]:
# STEP 1 — Hapus Duplikat
df.drop_duplicates(inplace=True)
print('Setelah hapus duplikat:', df.shape)

Setelah hapus duplikat: (130, 7)


In [15]:
# STEP 2 — Normalisasi String
df['kota'] = df['kota'].str.strip().str.title()
df['kondisi'] = df['kondisi'].str.strip().str.lower()
print('Normalisasi teks selesai.')

Normalisasi teks selesai.


In [16]:
# STEP 3 — Imputasi Missing Values
df['luas_m2'] = df['luas_m2'].fillna(df['luas_m2'].median())
df['harga_juta'] = df['harga_juta'].fillna(df['harga_juta'].median())
# Pakai [0] karena .mode() menghasilkan Series, kita ambil nilai pertamanya
df['kamar'] = df['kamar'].fillna(df['kamar'].mode()[0])
print('Data kosong telah diisi.')

Data kosong telah diisi.


In [19]:
# STEP 4 — Tangani Outlier (IQR Fence)
for col in ['harga_juta', 'luas_m2', 'tahun_bangun']:
    Q1, Q3 = df[col].quantile([0.25, 0.75])
    IQR = Q3 - Q1
    df[col] = df[col].clip(Q1-1.5*IQR, Q3+1.5*IQR)
print('Penanganan outlier selesai.')

Penanganan outlier selesai.


In [20]:
# STEP 5 — Validasi & Ekspor
assert df.isnull().sum().sum() == 0, 'Masih ada missing!'
assert df.duplicated().sum() == 0, 'Masih ada duplikat!'

print('Shape akhir:', df.shape)
df.to_csv('housing_clean.csv', index=False)
print('Dataset bersih tersimpan dengan nama: housing_clean.csv')

Shape akhir: (130, 7)
Dataset bersih tersimpan dengan nama: housing_clean.csv


### Kesimpulan

**Apa yang dipelajari:**
Pada Pertemuan ini yang dipelajari adalah pada alur kerja pembersihan data (*data cleaning pipeline*) menggunakan pandas, mulai dari penanganan nilai hilang, penghapusan duplikat, hingga normalisasi string dan penanganan outlier menggunakan metode IQR.

**Temuan Utama:**
- Dataset awal memiliki 130 baris dengan beberapa kolom yang memiliki nilai kosong signifikan (terutama `luas_m2` dan `harga_juta`).
- **Outlier terdeteksi** pada kolom `harga_juta`, `luas_m2`, dan `tahun_bangun`, yang kemudian berhasil ditangani menggunakan metode *clipping* (batas bawah dan atas).

**Keterbatasan & Pertanyaan:**
- Kolom 'kota' masih memiliki variasi penulisan (seperti 'Jogja' vs 'Yogyakarta') yang mungkin memerlukan pemetaan manual lebih lanjut di masa mendatang.